### PR TITLE
[DataFusion] -  Add show and show_limit function for DataFrame

### PR DESCRIPTION
**Pull request by francis-du**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/937

 # Rationale for this change

Collect the query results for the user in the show function and print the results to the console. 

If the user wants to preview the results, they only needs to call the `show` function.

# What changes are included in this PR?

Add `show` function implementation for DataFrame.

# Are there any user-facing changes?

Users can directly print the query results by calling the show function.

eg: 

```rust
async fn main() -> Result<()> {
  let mut ctx = ExecutionContext::new();
  let df = ctx.read_csv("tests/example.csv", CsvReadOptions::new())?;
  df.show().await?;
  Ok(())
}
```

+---+---+---+
| a | b | c |
+---+---+---+
| 1 | 2 | 3 |
+---+---+---+

### CODE REVIEW

File: datafusion/src/dataframe.rs
Comment (andygrove) [inline datafusion/src/dataframe.rs:239]: It would be nice if we could have similar behavior to Spark, where the default for `show` is to just show the first 20 rows.

```scala
def show(numRows: Int): Unit = show(numRows, truncate = true)

def show(): Unit = show(20)
```

File: pre-commit.sh
Comment (andygrove) [inline pre-commit.sh:23]: Are these pre-commit changes intentional here? This seems unrelated to adding the `show` function.

Comment (francis-du) [reply pre-commit.sh:23]: Yes, I found that pre-commit does not work, so I changed it, do I need to submit another PR?

File: datafusion/src/dataframe.rs
Comment (houqp) [inline datafusion/src/dataframe.rs:228]: We could add `no_run` here and remove the code comment there, so the doc string can still be tested to make sure it always compiles.

File: datafusion/src/dataframe.rs
Comment (francis-du) [reply datafusion/src/dataframe.rs:239]: DataFusion df has limit function to limit the number of rows, so I think it is just to add the default 20 rows, what do you think?

File: datafusion/src/dataframe.rs
Comment (francis-du) [reply datafusion/src/dataframe.rs:228]: I copied the code comment of the previous function, do I need to add it here?

Comment (houqp) [reply datafusion/src/dataframe.rs:228]: ha, ok, never mind then :)

File: datafusion/src/dataframe.rs
Comment (alamb) [reply datafusion/src/dataframe.rs:239]: I think adding a call to `df.limit(20)` prior to showing the output would work well

File: pre-commit.sh
Comment (alamb) [reply pre-commit.sh:23]: I would recommend a separate PR (mostly so we can tag the original author of the pre-commit and have a discussion about what to do there without holding up this PR)

**Review by alamb — APPROVED** [review]
I think this looks great -- thank you @francis-du !

File: pre-commit.sh
Comment (francis-du) [reply pre-commit.sh:23]: OK, I will open an issue and submit a PR to fix this.

File: datafusion/src/dataframe.rs
Comment (francis-du) [reply datafusion/src/dataframe.rs:239]: I think it can be inferred from the logical plan whether there is a limit number. If there is no limit, set it to the default value of 20. If there is, then the default limit value is not set.

Comment (alamb) [reply datafusion/src/dataframe.rs:239]: > This is an implementation:

I agree that looks perfect

Comment (francis-du) [reply datafusion/src/dataframe.rs:239]: OK, i pushed it, please review this changes.

File: pre-commit.sh
Comment (francis-du) [reply pre-commit.sh:23]: Just run `ln -s  ../../rust/pre-commit.sh .git/hooks/pre-commit` and exec `git commit`

**Comment by francis-du** [comment]
> I think this looks great -- thank you @francis-du !

By the way,should I need to add show function in example code?

File: datafusion/src/dataframe.rs
Comment (andygrove) [reply datafusion/src/dataframe.rs:239]: Sorry to be a pain but this seems confusing to me and maybe it would be better to revert to your original code and document that the user can limit output by using `df.limit(20).show()`. Alternatively, we could add a `show_limit(limit: usize)` alternate method where the user can specify how many rows they would like.

The issue I have with this is that it shows 20 rows by default and if the user wants more then they need to add a limit, which is counterintuitive because limit normally reduces the number of rows. Also, this code only works if the final operator is a limit, so it won't work constantly if the limit is wrapped in a sort, for example.

Comment (francis-du) [reply datafusion/src/dataframe.rs:239]: I greet, I will rewrite.

**Comment by loic-sharma** [comment]
The main README may be a good candidate here: https://github.com/apache/arrow-datafusion#example-usage

File: datafusion/src/dataframe.rs
Comment (alamb) [reply datafusion/src/dataframe.rs:239]: I defer to @andygrove

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @francis-du

File: README.md
Comment (alamb) [inline README.md:107]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @francis-du ! 

I enabled the CI checks to run and once they have passed I think this is good to me.

**Comment by francis-du** [comment]
> Thank you @francis-du ! 
> 
> I enabled the CI checks to run and once they have passed I think this is good to me. 
> 
> 

Thanks.

**Pull request merged**
